misk/src/main/kotlin/misk/web/exceptions/ExceptionHandlingInterceptor.kt

```diff
@@ -37,13 +45,67 @@ class ExceptionHandlingInterceptor(
     } catch (th: Throwable) {
       val response = toResponse(th)
       chain.httpCall.statusCode = response.statusCode
-      chain.httpCall.takeResponseBody()?.use { sink ->
-        chain.httpCall.addResponseHeaders(response.headers)
-        (response.body as ResponseBody).writeTo(sink)
+      if (chain.httpCall.dispatchMechanism == DispatchMechanism.GRPC) {
+        sendGrpcFailure(chain.httpCall, response)
+      } else {
+        sendHttpFailure(chain.httpCall, response)
       }
     }
   }
 
+  private fun sendHttpFailure(httpCall: HttpCall, response: Response<*>) {
+    httpCall.takeResponseBody()?.use { sink ->
+      httpCall.addResponseHeaders(response.headers)
+      (response.body as ResponseBody).writeTo(sink)
+    }
+  }
+
+  /**
+   * Borrow behavior from [GrpcFeatureBinding] to send a gRPC error with an HTTP 200 status code.
+   * This is weird but it's how gRPC clients work.
+   *
+   * One thing to note is for our metrics we want to pretend that the HTTP code is what we sent.
+   * Otherwise gRPC requests that crashed and yielded an HTTP 200 code will confuse operators.
+   */
```

Comment:
I see that this will reduce disruption with metrics, but I feel like having two sets of HTTP status codes is confusing in the long run, and it adds a layer of indirection with gRPC statuses. Why not have a metric for different gRPC statuses instead?

To provide a good migration path, we can probably emit both HTTP and gRPC status code metrics, then turn off HTTP status code metrics for gRPC endpoints later.

Reply:
Suggest new metrics that look like these, but no real need to make them identical I think? https://github.com/grpc-ecosystem/java-grpc-prometheus

Reply:
(fwiw, this is fine to keep as is in this PR)

Reply:
gRPC is always 200 OK, even when it’s not okay, and that is awful and a guarantee that there will be confusion. 

My preference is that we try to hide the insincere 200 OKs in implementation details. Metrics and alerting and even just developer expectations are pretty invested in 200/400/500 meaning different things. The 200 is a lie, and we want to quietly send it over the Wire without spreading that lie to other systems.

I think trying to get all those downstream tools to have independent HTTP and gRPC modes only spreads the confusion.

Reply:
I get that gRPC errors being HTTP 200 is confusing. Thinking about it more, my concerns are a) [HTTP 4xx/5xx and gRPC have a non-1:1 mapping with different semantics](https://grpc.github.io/grpc/core/md_doc_http-grpc-status-mapping.html), and b) hiding the HTTP and gRPC status makes it a bit more difficult to trace requests when debugging (e.g. the server said 502 in its logs, but the client got a 200).

I see what you mean with metrics though, and am ok with having a mapping for metrics purposes.

Reply:
I agree we should emit some gRPC-specific metrics to improve observability, but I think we can do that as a follow-up PR (#1992). This PR strictly improves the situation from "Misk counts 4xx/5xx metrics but breaks gRPC clients" to "Misk counts 4xx/5xx metrics and doesn't break gRPC clients". The third option, of "Misk fails to count 4xx/5xx metrics for gRPC but clients get the right error responses" seems strictly worse than either of those to me.